src/components/Proposal/Create/SchemeForms/CreateGenericMultiCallProposal.tsx

```diff
@@ -0,0 +1,502 @@
+        },
+      ],
+    });
+    this.state = {
+      loading: false,
+      tags: this.initialFormValues.tags,
+      addContractStatus: { error: "", message: "" },
+      whitelistedContracts: this.props.whitelistedContracts?.map(contract => { return contract.toLowerCase(); }) ?? [],
+      userContracts: [],
+    };
+  }
+
+  public async handleSubmit(formValues: IFormValues, { setSubmitting }: any): Promise<void> {
+    if (!await enableWalletProvider({ showNotification: this.props.showNotification })) { return; }
+
+    const contractsToCall = [];
+    const callsData = [];
+    const values = [];
+
+    for (const contract of formValues.contracts) {
+      contractsToCall.push(contract.address);
+      callsData.push(contract.callData);
+      values.push(contract.value);
+    }
+
+    const proposalValues = {
+      title: formValues.title,
+      description: formValues.description,
+      contractsToCall: contractsToCall,
+      callsData: callsData,
+      values: values,
+      url: formValues.url,
+      dao: this.props.daoAvatarAddress,
+      scheme: this.props.scheme.address,
+      tags: this.state.tags,
+    };
+
+    setSubmitting(false);
+    await this.props.createProposal(proposalValues);
+
+    Analytics.track("Submit Proposal", {
+      "DAO Address": this.props.daoAvatarAddress,
+      "Proposal Title": formValues.title,
+      "Scheme Address": this.props.scheme.address,
+      "Scheme Name": this.props.scheme.name,
+    });
+
+    this.props.handleClose();
+  }
+
+  // Exports data from form to a shareable url.
+  public exportFormValues(values: IFormValues) {
+    exportUrl({ ...values, ...this.state });
+    this.props.showNotification(NotificationStatus.Success, "Exportable url is now in clipboard :)");
+  }
+
+  private onTagsChange = (tags: any[]): void => {
+    this.setState({ tags });
+  }
+
+  /**
+ * Given a contract address, checks whether it's valid, not exists in the current contract list and that the contract is verified with valid ABI data and write methods.
+ * If all checks are okay, pushes the contract address to the contract lists, otherwise returns an appropriate message.
+ * @param {string} contractToCall
+ */
+  private verifyContract = async (contractToCall: string) => {
+    const addContractStatus: IAddContractStatus = {
+      error: "",
+      message: "",
+    };
+
+    if (contractToCall === "") {
+      addContractStatus.error = "";
+      addContractStatus.message = "";
+    } else if (!isAddress(contractToCall)) {
+      addContractStatus.error = "NOT_VALID_ADDRESS";
+      addContractStatus.message = "Please enter a valid address";
+    } else if (this.state.whitelistedContracts.includes(contractToCall) || this.state.userContracts.includes(contractToCall)) {
+      addContractStatus.error = "CONTRACT_EXIST";
+      addContractStatus.message = "Contract already exist!";
+    } else {
+      this.setState({ loading: true });
+      const abiData = await getABIByContract(contractToCall);
```

Comment:
is this not called twice ? one here for verify and other time at https://github.com/daostack/alchemy/pull/2197/files#diff-ea7dd2c9cc24fa59c55016881a6d0bddcc07ee4acf441f88989a3c32ac604fd3R196

Reply:
It's necessary because here we call it to check if this contract is verified and has write methods.
In line 196 we call each time we change a contract. I don't want to call all the list in advance because when you have a whitelist the user may not use all the contracts in the whitelist.

Reply:
what do you mean by `each time we change a contract.` ? you mean each time user enter a new call ?

Reply:
No, each time the user selects different contract in the select picker.